### PR TITLE
Copy args inside Handler and Logic

### LINKED ISSUE #308: Trainer: copy config in Handler
Handler should not modify config dict given by user.

### PR DESCRIPTION
**Pull request by emcastillo**
Fixes #308 

In order to do the validations inside the classes too, this PR superseedes #291

### CODE REVIEW

**Comment by kmaehashi** [comment]
* Could you update constructors and code of Runtime/Logic/Handler to use `option` instead of `config`?
* Please keep backward compatibility with the previous version like #291.

File: pytorch_pfn_extras/engine.py [hunk 1]
```diff
@@ -138,33 +140,40 @@ def create_trainer(
...
+        handler_options (dict):
+            Options that are set to the handler object. See the documentation
+            of `ppe.handler.Handler` for details.
+        logic_options (dict):
+            Options that arte set to the logic object. When using the default
+            logic class, See the documentation of `ppe.handler.Logic`
+            for details.
+        handler_options (dict):
+            Options that are set to the runtime object. See the documentation
+            of `ppe.handler.Handler` for details.
```
Comment (kmaehashi) [inline pytorch_pfn_extras/engine.py:158]:
```suggestion
        handler_options:
            Options that are set to the handler object. See the documentation
            of `ppe.handler.Handler` for details.
        logic_options:
            Options that are set to the logic object. When using the default
            logic class, See the documentation of `ppe.handler.Logic`
            for details.
        runtime_options:
            Options that are set to the runtime object. See the documentation
            of `ppe.handler.PyTorchRuntime` for details.
```

**Comment by emcastillo** [comment]
> Please keep backward compatibility with the previous version like Separate options argument into logic_options and runtime_options #291.

This is hard because if we pass an unified dict of logic/handler options then checking if all the options were not consumed in either of the class is going to be difficult.

File: pytorch_pfn_extras/engine.py (hunk 1 above)
Comment (emcastillo) [reply pytorch_pfn_extras/engine.py:158]: All the above params in the doc string have their type specified, should we remove all?

Comment (kmaehashi) [reply pytorch_pfn_extras/engine.py:158]: Updated suggestion, let's keep it as-is and discuss in #314 about that :smile:

**Comment by emcastillo** [comment]
PTAL!

**Comment by emcastillo** [comment]
rebased and removed types from docstring

**Pull request merged**